### PR TITLE
Fixed #260 & changed getUserAgent() return type

### LINKED ISSUE #260: Required dispatch_sync to load constants for RNDeviceInfo. This may lead to deadlocks
I'm getting this warning on 0.49.5
`Required dispatch_sync to load constants for RNDeviceInfo. This may lead to deadlocks`

Everything seems to function fine, but I'm wondering what does this exactly mean?

### PR DESCRIPTION
**Pull request by psegalen**
## Description

Fixes issue #260

/!\ Fix requires to change `getUserAgent()` return type (it was a `string` now it's a `Promise<string>`)

## Compatibility

| OS      | Implemented |
| ------- | :---------: |
| iOS     |    ✅     |
| Android |    ✅     |
| Windows |    ❌     |

## Checklist

* [X] I have tested this on a device/simulator for each compatible OS
* [X] I added the documentation in `README.md`.
* [ ] I mentionned this change in `CHANGELOG.md`.
> I'll let @machour modify the changelog because I don't know how he wants to inform everyone about the breaking change

### CODE REVIEW

File: ios/RNDeviceInfo/RNDeviceInfo.m [hunk 1]
```diff
@@ -314,4 +306,22 @@ - (NSDictionary *)constantsToExport
...
+RCT_EXPORT_METHOD(getUserAgent:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject)
+{
+    dispatch_async(dispatch_get_main_queue(), ^{
+        if (_webView == nil) {
+            _webView = [[WKWebView alloc] initWithFrame:CGRectZero];
+        }
+        [_webView evaluateJavaScript:@"navigator.userAgent" completionHandler:^(id result, NSError * error) {
+            if (error == nil) {
+                if (result != nil) {
+                    resolve([NSString stringWithFormat:@"%@", result]);
```
Comment (maxkomarychev) [inline ios/RNDeviceInfo/RNDeviceInfo.m]: [docs](https://developer.apple.com/documentation/webkit/wkwebview/1415017-evaluatejavascript) say:

> [...] The completion handler always runs on the main thread.

perhaps it's better to resolve this not on main thread?
can we utilize `self.methodQueue`?

File: ios/RNDeviceInfo/RNDeviceInfo.m [hunk 2]
```diff
@@ -314,4 +306,22 @@ - (NSDictionary *)constantsToExport
...
+                }
+            } else {
+                reject(@"evaluate_js_failed", [NSString stringWithFormat:@"evaluateJavaScript error : %@", error.localizedDescription], error);
```
Comment (maxkomarychev) [inline ios/RNDeviceInfo/RNDeviceInfo.m]: same as above

File: ios/RNDeviceInfo/RNDeviceInfo.m (hunk 1 above)
Comment (maxkomarychev) [inline ios/RNDeviceInfo/RNDeviceInfo.m]: is `result` not a string?

File: ios/RNDeviceInfo/RNDeviceInfo.m [hunk 3]
```diff
@@ -314,4 +306,22 @@ - (NSDictionary *)constantsToExport
...
 
+RCT_EXPORT_METHOD(getUserAgent:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject)
+{
+    dispatch_async(dispatch_get_main_queue(), ^{
+        if (_webView == nil) {
+            _webView = [[WKWebView alloc] initWithFrame:CGRectZero];
+        }
+        [_webView evaluateJavaScript:@"navigator.userAgent" completionHandler:^(id result, NSError * error) {
+            if (error == nil) {
+                if (result != nil) {
```
Comment (maxkomarychev) [inline ios/RNDeviceInfo/RNDeviceInfo.m]: This looks like promise will never be resolved with if `result == nil`.
Can we do better like:

```
if (result) {
  resolve([NSString stringWithFormat:@"%@", result]); // still not sure if this is necessary but I guess docs don't guarantee type of the result...
} else {
  resolve(nil)
}
```
makes sense?

File: ios/RNDeviceInfo/RNDeviceInfo.m (hunk 1 above)
Comment (psegalen) [reply ios/RNDeviceInfo/RNDeviceInfo.m]: Implicit cast can be dangerous, I'll add a Type check

Comment (psegalen) [reply ios/RNDeviceInfo/RNDeviceInfo.m]: I didn't create a `methodQueue` only for that so I dispatch on global priority default queue like shown in https://facebook.github.io/react-native/docs/native-modules-ios.html#threading

File: ios/RNDeviceInfo/RNDeviceInfo.m (hunk 3 above)
Comment (maxkomarychev) [reply ios/RNDeviceInfo/RNDeviceInfo.m]: I saw `resolve(nil)` in many libs, but that's not too critical :)

**Comment by ishabo** [comment]
@psegalen can you please resolve the conflicts?

**Comment by psegalen** [comment]
@ishabo done, sorry for the lag!

**Comment by SudoPlz** [comment]
Any news with that?

**Comment by mikehardy** [comment]
Linking this to the #307 v2.0 issue where the main goal will be to Promise-ify the API, since this appears to be tested code.

Worth noting that iOS doesn't crash with master right now, using at least react-native 0.58+ (I did not test 0.57)

**Comment by rajivshah3** [comment]
@psegalen would you mind resolving the conflicts again? Hopefully a maintainer will be able to review this PR soon, as it should improve launch performance on apps that use this library

**Comment by mikehardy** [comment]
The PR is reviewed, it is the direction we will go (so you can feel comfortable using this style), but it is a large breaking change so it hasn't happened yet

File: web/index.js [hunk 4]
```diff
@@ -20,7 +20,7 @@ module.exports = {
   buildNumber: 0,
   appVersion: 0,
   deviceName: '',
-  userAgent: window.navigator.userAgent,
+  userAgent: () => new Promise((resolve, reject) => resolve(window.navigator.userAgent)),
```
Comment (LinusU) [inline web/index.js]: This could be just `() => Promise.resolve(window.navigator.userAgent)` which avoids creating an extra closure :)

**Comment by mikehardy** [comment]
I'm not too bothered by the conflicts, and this is still the direction I want to go. In the meantime the whole AndroidX/jetifier thing happened though which unfortunately absorbed the time I would have had for the Promise API switch over. It will happen eventually though, and I appreciate the patience. Also if anyone else just couldn't take it anymore and converted everything to promises in a PR that'd be a good way to get it done ;-)

**Comment by ldco2016** [comment]
Any solution to this issue that got locked?
https://github.com/react-native-community/react-native-device-info/issues/260

I get a yellow warning with `Required dispatch_sync to load constants for RNDeviceInfo. This may lead to deadlocks.`

I am using `"react-native-device-info": "2.2.2",` with `"react-native": "0.60.4",`

**Comment by mikehardy** [comment]
No deep solution, currently just ignore. "may lead to deadlocks", but does not in practice. A future version of react-native-device-info could solve it but it requires the conversion to promise etc that has been pending but was delayed by me focusing on the whole RN60/jetifier thing, and no one else submitting PRs 🤷‍♂

**Comment by ldco2016** [comment]
@mikehardy , thanks for clarifying and getting back to me. I understand about the whole jetifier androidx thing, was upgrading a project when all that happened.

**Comment by mikehardy** [comment]
Obsoleted by #757 as I shape up a v3 but this work (esp the java part and docs and js impl) will be informed by it - thank you, and thanks for your patience

**Pull request closed without merging**